walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,66 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.Set;
+import java.util.HashSet;
+import java.util.List;
+import java.util.ArrayList;
+import java.util.Arrays;
+
+import com.google.sps.Event;
+import com.google.sps.TimeRange;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Collection<String> mandatoryAttendees = request.getAttendees();
+    Collection<String> optionalAttendees = request.getOptionalAttendees();
+    Set<String> allAttendees = new HashSet<>(mandatoryAttendees);
+    allAttendees.addAll(optionalAttendees);
+    
+    Collection<TimeRange> ranges = getTimeRanges(events, allAttendees, request.getDuration());
+    if (ranges.size() > 0) {
+      return ranges;
+    }
+    if (mandatoryAttendees.size() > 0) {
+      return getTimeRanges(events, mandatoryAttendees, request.getDuration());
+    }
+    return Arrays.asList();
+  }
+
+  public Collection<TimeRange> getTimeRanges(Collection<Event> events, Collection<String> attendees, long duration) {
+    Comparator<Event> comparator = new Comparator<Event>() {
+      @Override
+      public int compare(final Event e1, final Event e2) {
+        return TimeRange.ORDER_BY_START.compare(e1.getWhen(), e2.getWhen());
+      }
+    };
+    boolean rejectedEvent = false;
+    List<Event> l = new ArrayList(events);
+    Collections.sort(l, comparator);
+    Collection<TimeRange> ranges = new ArrayList<TimeRange>();
+    int start = TimeRange.START_OF_DAY;
+    for (Event e : l) {
```

Comment:
this code block would benefit from a comment that provide a high level overview of what you're trying to achieve and how. this saves the reader some effort in figuring out what type of algorithm you are attempting.

Reply:
Added.